fastly.go

```diff
@@ -14,41 +13,47 @@ import (
 
 func getServiceNames(token string, serviceIDs []string, logger log.Logger) map[string]string {
 	serviceNames := map[string]string{}
-	for _, serviceID := range serviceIDs {
-		serviceNames[serviceID] = getServiceName(token, serviceID, log.With(logger, "service_id", serviceID))
-	}
-	return serviceNames
-}
 
-func getServiceName(token, serviceID string, logger log.Logger) string {
-	req, err := http.NewRequest("GET", fmt.Sprintf("https://api.fastly.com/service/%s/details", url.QueryEscape(serviceID)), nil)
+	req, err := http.NewRequest("GET", "https://api.fastly.com/service", nil)
 	if err != nil {
 		level.Error(logger).Log("err", err)
-		return serviceID
+		return serviceNames
 	}
 
 	req.Header.Set("Fastly-Key", token)
 	req.Header.Set("Accept", "application/json")
 	resp, err := http.DefaultClient.Do(req)
 	if err != nil {
 		level.Error(logger).Log("err", err)
-		return serviceID
+		return serviceNames
 	}
 	defer resp.Body.Close()
 
-	var response struct {
-		Name string `json:"name"`
-	}
+	var response serviceResponse
+
 	if err := json.NewDecoder(resp.Body).Decode(&response); err != nil {
 		level.Error(logger).Log("err", err)
-		return serviceID
+		return serviceNames
+	}
+
+	for _, entry := range response {
+		serviceNames[entry.ID] = entry.Name
 	}
-	if response.Name == "" {
-		level.Error(logger).Log("err", "Fastly returned blank service name")
-		return serviceID
+
+	if len(serviceIDs) > 0 {
+		limitedServiceNames := map[string]string{}
```

Comment:
Maybe it would be nicer if the -service flags didn't flag.Var to a string slice, but rather directly to a map[string]string, where the key is service ID (provided by user) and the value is initially the empty string but could be looked up by this function?  If no -service is given, then this function could do a first pass to look up all service IDs via the API, and kill the process if it doesn't work. I haven't coded it through, but my intuition is that may make things easier—what do you think?

Reply:
I'm ambivalent on this one.  I didn't touch the -service flag construction as it works well enough and doesn't impact any features I would like to add ;)